File: lc_soln/0074_search_a_2d_matrix/solution.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int nc = matrix[0].size();

        int t = 0;
        int b = matrix.size() - 1;
        while (t <= b) {
            int row = t + (b - t) / 2;
            if (matrix[row][0] <= target && target <= matrix[row][nc - 1]) {
                int l = 0;
                int r = nc - 1;
                while (l <= r) {
                    int mid = l + (r - l) / 2;
                    if (matrix[row][mid] == target) {
                        return true;
                    }
                    if (matrix[row][mid] < target) {
                        l = mid + 1;
                    } else {
                        r = mid - 1;
                    }
                }
            }
            if (matrix[row][nc - 1] < target) {
                t = row + 1;
            } else {
                b = row - 1;
            }
        }
        return false;
    }
};

int main() {
    return 0;
}

File: lc_soln/0074_search_a_2d_matrix/solution.rs
struct Solution;

impl Solution {
    pub fn search_matrix(matrix: Vec<Vec<i32>>, target: i32) -> bool {
        let nr = matrix.len();
        let nc = matrix[0].len();

        let mut t = 0i32;
        let mut b = (nr - 1) as i32;

        let mut row = nr;
        while t <= b {
            let mid = (t + (b - t) / 2) as usize;
            if matrix[mid][0] <= target && target <= matrix[mid][nc - 1] {
                row = mid;
                break;
            }
            if matrix[mid][nc - 1] < target {
                t = (mid + 1) as i32;
            } else {
                b = (mid - 1) as i32;
            }
        }

        if row == nr {
            return false;
        }

        let mut l = 0i32;
        let mut r = (nc - 1) as i32;
        while l <= r {
            let mid = (l + (r - l) / 2) as usize;
            if matrix[row][mid] == target {
                return true;
            }
            if matrix[row][mid] < target {
                l = (mid + 1) as i32;
            } else {
                r = (mid - 1) as i32;
            }
        }
        false
    }
}

fn main() {}
